core/src/main/java/com/linecorp/armeria/server/RouteBuilder.java

```diff
@@ -267,6 +279,102 @@ public RouteBuilder produces(Iterable<MediaType> produceTypes) {
+     *     <li>{@code some-param} which the request must contain a {@code some-param} parameter</li>
+     *     <li>{@code !some-param} which the request must not contain a {@code some-param} parameter</li>
+     * </ul>
+     *
+     * <p>Note that these predicates can be evaluated only with the query string of the request URI.
+     * Also note that each predicate will be evaluated with the decoded value of HTTP parameters,
+     * so do not use percent-encoded value in the predicate.
+     *
+     * @see Params
+     */
+    public RouteBuilder params(Iterable<String> paramPredicates) {
+        this.paramPredicates = ImmutableList.copyOf(requireNonNull(paramPredicates, "paramPredicates"));
+        return this;
+    }
+
+    /**
+     * Sets the {@link Route} to accept a request if it matches all the specified predicates for
+     * {@link HttpHeaders}. The predicate can be one of the following forms:
+     * <ul>
+     *     <li>{@code some-header=some-value} which the request must have a
+     *     {@code some-header: some-value} header</li>
+     *     <li>{@code some-header!=some-value} which the request must not have a
+     *     {@code some-header: some-value} header</li>
+     *     <li>{@code some-header} which the request must contain a {@code some-header} header</li>
+     *     <li>{@code !some-header} which the request must not contain a {@code some-header} header</li>
+     * </ul>
+     *
+     * @see Headers
+     */
+    public final RouteBuilder headers(String... headerPredicates) {
+        return headers(ImmutableList.copyOf(requireNonNull(headerPredicates, "headerPredicates")));
+    }
+
+    /**
+     * Sets the {@link Route} to accept a request if it matches all the specified predicates for
+     * {@link HttpHeaders}. The predicate can be one of the following forms:
+     * <ul>
+     *     <li>{@code some-header=some-value} which the request must have a
+     *     {@code some-header: some-value} header</li>
+     *     <li>{@code some-header!=some-value} which the request must not have a
+     *     {@code some-header: some-value} an header</li>
+     *     <li>{@code some-header} which the request must contain a {@code some-header} header</li>
+     *     <li>{@code !some-header} which the request must not contain a {@code some-header} header</li>
+     * </ul>
+     *
+     * @see Headers
+     */
+    public RouteBuilder headers(Iterable<String> headerPredicates) {
+        this.headerPredicates = ImmutableList.copyOf(requireNonNull(headerPredicates, "headerPredicates"));
+        return this;
+    }
+
+    /**
+     * TODO(hygntack) Is this useful?.
+     *
+     * @param id the identifier of the specified {@code headersPredicate} which is used for logging or
+     *           metric collection
+     * @param headersPredicate the predicates to be used when accepting a request by a {@link Route}
+     */
+    public RouteBuilder headers(String id, Predicate<HttpHeaders> headersPredicate) {
```

Comment:
I think it will be useful, especially for ourselves; we could re-implement `consumes()` and `produces()` using this feature, and I guess a user may have a similar use case.

However, I'm not sure if it's a good idea to let a user specify `id` here. I think we should solve this problem at higher-level, like letting a user assign an `id` to a `Route`.

Reply:
Yeah, I agree with you. Would it be better to address it with follow-up PR?